scheduler/scheduler.go

```diff
@@ -139,63 +140,67 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
 			continue
 		}
 
+		if ctx.Err() != nil {
+			return nil, errors.New("scheduling run canceled.")
+		}
+
+		d, err := distro.FindOne(distro.ById(distroId))
+		if err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"distro":  distroId,
+				"runner":  RunnerName,
+				"message": "failed to find distro",
+			}))
+			return nil, err
+		}
+
 		hostsSpawnedPerDistro[distroId] = make([]host.Host, 0, numHostsToSpawn)
-		for i := 0; i < numHostsToSpawn; i++ {
-			if ctx.Err() != nil {
-				return nil, errors.New("scheduling run canceled.")
+		// if distro can have containers, check if there are enough parents to hold
+		// new containers
+		if d.MaxContainers > 0 {
+			currentParents, err := host.FindAllRunningParents()
+			if err != nil {
+				return nil, err
 			}
-
-			d, err := distro.FindOne(distro.ById(distroId))
+			existingContainers, err := host.FindAllRunningContainers()
 			if err != nil {
-				grip.Error(message.WrapError(err, message.Fields{
-					"distro":  distroId,
-					"runner":  RunnerName,
-					"message": "failed to find distro",
-				}))
-				continue
+				return nil, err
 			}
 
-			allDistroHosts, err := host.Find(host.ByDistroId(distroId))
-			if err != nil {
-				grip.Error(message.WrapError(err, message.Fields{
-					"distro":  distroId,
-					"runner":  RunnerName,
-					"message": "problem getting hosts for distro",
-				}))
+			numNewParents := calcNewParentsNeeded(len(currentParents),
+				len(existingContainers), numHostsToSpawn, d)
 
-				continue
-			}
+			numNewParents, numHostsToSpawn = parentCapacity(d, numNewParents, len(currentParents), len(existingContainers), numHostsToSpawn)
```

Comment:
Since you calculate `numHostsToSpawn` from `containerCapacity`, I think `parentCapacity` doesn't need to also calculate it.